tools/gfx/vulkan/render-vk.cpp

```diff
@@ -879,322 +676,2084 @@ class VKDevice : public GraphicsAPIRenderer
+                auto& imageInfo = ((VkDescriptorImageInfo*)infos)[i];
+                imageInfo.imageView = texture->m_view;
+                imageInfo.imageLayout = texture->m_layout;
+                imageInfo.sampler = sampler->m_sampler;
+            }
+            bindingState->descriptorSetWrites.add(write);
+        }
+
+        static void writeTextureDescriptor(
+            RootBindingState* bindingState,
+            BindingOffset offset,
+            VkDescriptorType descriptorType,
+            ArrayView<RefPtr<ResourceViewImpl>> resourceViews)
+        {
+            auto descriptorSet = bindingState->descriptorSets[offset.descriptorSetIndexOffset];
+            VkWriteDescriptorSet write = {};
+            write.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+            write.descriptorCount = (uint32_t)resourceViews.getCount();
+            write.descriptorType = descriptorType;
+            write.dstArrayElement = 0;
+            write.dstBinding = offset.descriptorRangeOffset;
+            write.dstSet = descriptorSet;
+            auto infos = bindingState->descriptorInfos.reserveRange(write.descriptorCount);
+            write.pImageInfo = (VkDescriptorImageInfo*)infos;
+            for (Index i = 0; i < resourceViews.getCount(); i++)
+            {
+                auto texture = static_cast<TextureResourceViewImpl*>(resourceViews[i].Ptr());
+                auto& imageInfo = ((VkDescriptorImageInfo*)infos)[i];
+                imageInfo.imageView = texture->m_view;
+                imageInfo.imageLayout = texture->m_layout;
+                imageInfo.sampler = 0;
+            }
+            bindingState->descriptorSetWrites.add(write);
+        }
+
+        static void writeSamplerDescriptor(
+            RootBindingState* bindingState,
+            BindingOffset offset,
+            VkDescriptorType descriptorType,
+            ArrayView<RefPtr<SamplerStateImpl>> samplers)
+        {
+            auto descriptorSet = bindingState->descriptorSets[offset.descriptorSetIndexOffset];
+            VkWriteDescriptorSet write = {};
+            write.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+            write.descriptorCount = (uint32_t)samplers.getCount();
+            write.descriptorType = descriptorType;
+            write.dstArrayElement = 0;
+            write.dstBinding = offset.descriptorRangeOffset;
+            write.dstSet = descriptorSet;
+            auto infos = bindingState->descriptorInfos.reserveRange(write.descriptorCount);
+            write.pImageInfo = (VkDescriptorImageInfo*)infos;
+            for (Index i = 0; i < samplers.getCount(); i++)
+            {
+                auto& imageInfo = ((VkDescriptorImageInfo*)infos)[i];
+                imageInfo.imageView = 0;
+                imageInfo.imageLayout = VK_IMAGE_LAYOUT_GENERAL;
+                imageInfo.sampler = samplers[i]->m_sampler;
+            }
+            bindingState->descriptorSetWrites.add(write);
+        }
```

Comment:
Hypothetically, the shader object *layout* could pre-compute all these `VkWriteDescriptorSet` and related objects into a single compact buffer, and then a particular object would just need to use a scratch-allocated copy that it then patches a few pointers into. (One challenge here is that the VK API loves its little structure types, but then misses a lot of the potential benefit of re-using those structures once written by mixing up pointer and non-pointer data all over the place)